### ISSUE TITLE
model predict t-1 value

### ISSUE THREAD

**Issue by Picus303** [facebook/prophet#1875]
Hi,
I trained a model using a dataset that I previously splited to be able to test it, but when I try to get the future values each value predicted is the t-1 value (exemple: I got the value which is suppose to be for 19:45 at 19:46).
After that I tryed to shift my data but nothing changed.
Plus, I have additionnal regressors and I'm gonna make predictions on one period with real time data so I can't play to much with the dates.
Is it normal?

**Comment by bletham** [comment]
I don't really understand what the issue is, could you be more specific about what functions you're using and what dataframes you're looking at? In particular, if you could post a minimal code example that produces the issue that'd be great. Generally, if you look at `m.history` (Py) or `m$history` (R) you'll see the history dataframe being used by the model, and can verify that that matches the training data that you passed in for fitting. And on the `fcst = m.predict(future)` side, it will produce predictions for whatever dates you pass in to it, so you should see that `fcst['ds']` and `future['ds']` are the same.

**Comment by Picus303** [comment]
I tryed everything you said so: `m.history` return the right dataframe and `fcst['ds']` and `future['ds']` are the same.
Here is exactly the code I'm using:

`import pandas as pd
df=pd.read_csv("temp.csv")

df['v']=df['v'].shift(1)
df['h']=df['h'].shift(1)
df['l']=df['l'].shift(1)
df.dropna(axis=0, inplace=True)

df['t']=pd.to_datetime(df['t'])
df=df[(df['t'].dt.hour<23) & (df['t'].dt.hour>9)]
df=df.iloc[:-13700]

df.columns=["ds", "v", "o", "y", "h", "l"]

test=df.iloc[-60000:]
df=df.iloc[:-60000]

from fbprophet import Prophet

model=Prophet()
model.add_regressor('v')
model.add_regressor('h')
model.add_regressor('l')
model.add_regressor('o')
model.fit(df)

df=pd.read_csv("temp.csv")
df['t']=pd.to_datetime(df['t'])
df=df.set_index('t')
df['v']=df['v'].shift(1)
df['h']=df['h'].shift(1)
df['l']=df['l'].shift(1)
df=pd.DataFrame(df.loc['2020-08-07 19:16:00':'2020-08-07 20:16:00'])
df.dropna(axis=0, inplace=True)

futur=model.make_future_dataframe(periods=100, freq='min')
futur=futur.merge(df, left_on='ds', right_index=True)
futur.dropna(axis=0, inplace=True)

forecast=model.predict(futur)`

And then I compare.

Also, here is the dataframe I'm using.
![Capture](https://user-images.githubusercontent.com/82946075/115606111-95d56c00-a2e3-11eb-902d-942e8664bfd7.PNG)

**Comment by bletham** [comment]
Could you explain what the issue is? I don't see anything in your code that looks incorrect, but I also don't understand what the issue is. What are you comparing and what is the discrepancy?

**Comment by Picus303** [comment]
Ok: I think that it will be easier with a picture:
![Capture](https://user-images.githubusercontent.com/82946075/116368857-0b23ce00-a809-11eb-9879-161b32a2d0ab.PNG)
The red line is my data, the blue line is my prediction. 
You can see that when I compare the values there is a one minute shift.
And it's the same after shifting my columns to try to obtain the right values.

Here is the code that I use for comparaison:
`
import pandas as pd

df=pd.read_csv("temp.csv") #same dataframe as for training
df['t']=pd.to_datetime(df['t'])
df=df.set_index('t')
df['v']=df['v'].shift(1)
df['l']=df['l'].shift(1)
df['h']=df['h'].shift(1)
df=pd.DataFrame(df.loc['2020-08-07 19:16:00':'2020-08-07 20:16:00'])
df.dropna(axis=0, inplace=True)

future=model.make_future_dataframe(periods=100, freq='min')

future=future.merge(df, left_on='ds', right_index=True)
future.dropna(axis=0, inplace=True)

forecast=model.predict(future)

df_comp=df['c']
forecast_comp=forecast.set_index('ds')['yhat']

import matplotlib.pyplot as plt
%matplotlib inline

plt.plot(df_comp.reset_index(drop=True), 'r')
plt.plot(forecast_comp.reset_index(drop=True), 'b')
`

**Comment by bletham** [comment]
I suspect something is going wrong in one of the merges but I'm not sure. Could you attach `temp.csv`? You might need to put it in a zip in order to attach it.

**Comment by Picus303** [comment]
Here is it:
[temp.csv](https://github.com/facebook/prophet/files/6392584/temp.csv)

**Comment by bletham** [comment]
Thanks, I was really struggling to figure out what was going on here but I see now.

This time series doesn't really have the type of structure that Prophet works best for - in particular, there is no seasonality, and not really a clear trend either. You have added extra regressors to the model. Extra regressors in Prophet are added as linear regressors to the model. Here the regressors have been shifted by one, so the model here is essentially
```
yhat(t) = trend(t) + seasonality(t) + beta1 * l(t-1) + beta2 * o(t-1) ...
```
where beta1, beta2, etc. are fit to the data. Note that the regressors are linear terms in the model. This means that y(t) must be directly proportional to v(t-1), etc.

Because there is very little trend and no seasonality, basically the model is
```
yhat(t) = beta1 * l(t-1) + beta2 * o(t-1) ...
```
If you plot y(t) vs. l(t), you will see that they are very highly correlated.  But we're not including l(t) in the model, we're including l(t-1). Now, l(t-1) is still pretty highly correlated with y(t), and so when we set yhat(t) = l(t-1), that's the best that Prophet can do because there is no other structure in the time series that it can use to try to predict y.

You can see this phenomenon very clearly if you look at
```
plt.plot(df2['y'].values, 'r')
plt.plot(df2['l'].values, 'b')
```
This looks basically the same as the plot that you made above, because `yhat` is basically the same as `l` since it is being used as a linear regressor and there is nothing else for it to use to predict `y`.

Does that make sense?

**Comment by Picus303** [comment]
Yes, I think. Thanks a lot!
I'll try something else to see if I can fix that with the informations you gave me.
By the way now that I understand the problem, if you think of a way to fix it you're welcome.
